compiler/rustc_borrowck/src/region_infer/mod.rs

```diff
@@ -2168,17 +2175,38 @@ impl<'tcx> RegionInferenceContext<'tcx> {
                 }
             }
 
+            // If an "outlives 'static" constraint was from use as a const or static, blame that.
+            if target_is_static
+                && blame_source
+                && let Some(old_best) = categorized_path.iter().min_by_key(|p| p.category)
```

Comment:
this feels odd to me, do you want to disable this if there's also one step in the path whose category is a `ConstraintCategory::Return` or would it be pattern to walk the path to detect whether any `ConstraintCategory::UseAsConst` exists?

Reply:
looking at it again, that special case shouldn't be there at all; it's just papering over a more fundamental issue. from what I can tell, the reason [tests/ui/consts/const-mut-refs/mut_ref_in_final.stderr](https://github.com/dianne/rust/blob/189b2f892e6d0809a77fc92fe1108a07d8de9be0/tests/ui/consts/const-mut-refs/mut_ref_in_final.stderr#L38) needs the special-casing is that `UnsafeCell<T>` is invariant in `T`, hence the outliving region of the `UseAsConst`/`UseAsStatic` constraint is in the same SCC as the target, hence it gets skipped. for a quick workaround, this could check if there's a constraint on the path with an invariant `variance_info` before removing any edges from the region graph. it's sort of justifiable, since back-edges from invariance would also need to be removed in order to get the properties I wanted from pruning the `'static: R` edges. that said, I think this should handle invariance properly if it can. leaving it sometimes-wrong feels pretty unsatisfying, especially if a proper fix might handle this whole issue cleanly. I'll have to look into it further.